kubernetes, helm: Update to the latest version of the operator

krnowak opened the pull request:
- generate a new version of the manifest (`v1beta2` custom version)
- generate persistent storage fields in the manifest
- generate environment variable fields in the manifest

baumanj committed:
Initial Expeditor configuration for updating CHANGELOG.md

Resolves https://github.com/habitat-sh/habitat/issues/4654

Signed-off-by: Jon Bauman <5906042+baumanj@users.noreply.github.com>

baumanj committed:
Remove deprecated 'maintainer_group' setting

Signed-off-by: Jon Bauman <5906042+baumanj@users.noreply.github.com>

christophermaier committed:
Update Documentation for 0.56.0 Release

Includes documentation on new remote-control functionality, as well as
a rewrite of the binding documentation.

Uses of `hab start <IDENT>` were replaced with `hab svc load <IDENT>`
in nearly all places, with additional documentation and caveating
around when to use `hab sup run <IDENT>` (tl;dr - rarely, if ever).

How to run Habitat on servers was pulled out into a separate page.

There are various other formatting and wording tweaks scattered
around, as well.

Signed-off-by: Christopher Maier <cmaier@chef.io>

krnowak committed:
pkg-export-k8s: Fix some typos and grammar

Service name and group name in service group are separated with a dot,
not with a colon.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
pkg-export-k8s: Generate bind json values a bit differently

That way we don't have to have a mutable binds variable.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
pkg-export-k8s: Generate v1beta2 custom version of habitat manifest

This only changes the version of the manifest the exporter will
generate, but does not add support for generating the new fields in
the manifest yet. The only difference with the previous version of the
generated manifest is that the Habitat operator will use Kubernetes
stateful sets instead of deployments as a way to deploy services in
the cluster.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
pkg-export-k8s: Move manifest rendering to a separate struct

We use handlebars renderer to generate a YAML document, so we need to
disable the escaping, because handlebars by default uses an
HTML-specific escaping.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
pkg-export-k8s: Add a quote helper to manifest generator

We will want to generate fields with arbitrary values in the YAML
document, so we need to put them inside quotes. For this to work, the
string inside the quotes needs to have backslashes and quotes
escaped. The new helper does exactly that - escapes the necessary
chars and wraps the string in quotes.

This will be needed for generating environment variables fields in
Kubernetes manifest.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
pkg-export-k8s: Generate environment variables fields in manifest

This adds a `--env` flag that can be used multiple times to add a
environment variable to the manifest.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
pkg-export-k8s: Generate persistent storage fields in manifest

This adds a `--storage` flag that can be used to specify a persistent
storage inside the manifest for the service.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
pkg-export-k8s: Expose PersistentStorage and QuoteHelper

Helm exporter will use both.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
pkg-export-helm: Avoid useless cloning

We don't need to take any values from inside the Option, we only need
to know if the option has any value in it. So just use as_ref() which
creates a new Option with a reference value, so it is not copied
unnecessarily.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
pkg-export-helm: Use an enumerating iterator instead of a counter

One variable less and it is doing the incrementation for us.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
pkg-export-helm: Mention another helm issue in deps generator

The first issue was closed in favor of the second one, but keep the
first one for context.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
pkg-export-helm: Fix a typo

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
pkg-export-helm: Bump a default operator version

It is the version that supports a new custom version with stateful
sets and persistent storage.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
pkg-export-helm: Handle persistent storage fields in manifest

The support is done through the three new values in values.yaml -
`storageSize`, `storagePath` and `storageClass`.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
pkg-export-helm: Use QuoteHelper to generate values

Values will contain environment variable values, which can be
arbitrary. Since it is a YAML document, we need to escape backslashes
and quotes therein and wrap the result in quotes.

Just use QuoteHelper from the kubernetes exporter to do the work.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
pkg-export-helm: Handle environment variable fields in manifest

The support is done through the multiple value pairs in values.yaml
(similar to binds) - `envName<NUM>` and `envValue<NUM>`.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

cnunciato committed:
Update Supervisor tutorials

This change updates the Supervisor-related tutorials on habitat.sh to bring them in line with the changes coming in the next release.

Signed-off-by: Christian Nunciato <chris@nunciato.org>

iaguis commented on components/pkg-export-kubernetes/src/hb.rs:
```diff
@@ -12,12 +12,52 @@
+            })?
+            .value()
+            .as_str()
+            .ok_or_else(|| RenderError::new("Expected the string parameter"))?;
```
Expected **a** string parameter

iaguis commented on components/pkg-export-kubernetes/src/cli.rs:
```diff
@@ -146,6 +146,20 @@ impl<'a, 'b> Cli<'a, 'b> {
+                            "Storage specification in form of <size>:<path>:<storage class name>. \
+                             <size> uses the same format as Kubernetes' size field (e.g. 10Gi). \
+                             <path> describes where the storage will be mounted. \
+                             <storage class name> is a name of a storage class \
```
is **the** name of **the** storage class

iaguis commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -52,6 +53,8 @@ pub struct Manifest {
     /// Any binds, as `ServiceBind` instances.
     pub binds: Vec<ServiceBind>,
+    /// Persistent storage information
+    pub storage: Option<PersistentStorage>,
```
Also here, should we use `persistent_storage` for consistency?

Along with similar changes in `cli.rs`.

Not a big deal.

iaguis commented on components/pkg-export-kubernetes/src/storage.rs:
```diff
@@ -0,0 +1,193 @@
+
+    fn valid_size(size: &str) -> bool {
+        #[derive(Clone, Copy, PartialEq)]
+        enum State {
```
State Machines :heart_eyes:

iaguis commented on components/pkg-export-helm/src/chart.rs:
```diff
@@ -137,6 +137,23 @@ impl<'a> Chart<'a> {
+            values.add_entry(&value_var, &envvar.value);
+
+            let json = json!({
+                "name": format!("{{{{.Values.{}}}}}", name_var),
```
braces, braces everywhere. Interesting escaping.

krnowak replied:
Yeah, use double bracket to get a bracket. Tricked me once, when I tried using `\{`.

iaguis commented on components/pkg-export-kubernetes/src/error.rs:
```diff
@@ -24,6 +24,15 @@ pub enum Error {
+                      <PATH> is an absolute path",
+           _0)]
+    InvalidStorage(String),
+    #[fail(display = "Invalid environment variable \"{}\", must in the form <NAME>=<VALUE>", _0)]
```
must **be** in the form

iaguis commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -51,6 +53,10 @@ pub struct Manifest {
+    /// Persistent storage information
+    pub storage: Option<PersistentStorage>,
+    /// Environment
+    pub environment: Vec<EnvVar>,
```
Since the name in the Habitat type is `env` this should also be `env`. I guess you use `environment` here because of some name clash?

Anyway, not a big deal.

iaguis commented on components/pkg-export-helm/src/chart.rs:
```diff
@@ -75,14 +75,25 @@ impl<'a> Chart<'a> {
+                .as_ref()
                 .map(|_| "{{.Values.ringSecretName}}"),
-        }),
+            "storage": manifest.storage
```
Same thing about storage and persistent_storage

iaguis reviewed (changes requested):
Some small comments. Otherwise (and taking into account I don't know much rust) looks good. Also, I haven't tested it :)

christophermaier committed:
Move binding docs to the right place

Turns out we had 2 files that were largely identical, only one of
which was actually linked to anything. The first time around, I chose
the wrong one of those to make my edits in.

Signed-off-by: Christopher Maier <cmaier@chef.io>

christophermaier committed:
Use DNS names for `--remote-sup` examples, because we can

Signed-off-by: Christopher Maier <cmaier@chef.io>

cnunciato committed:
Update the Supervisor demo to line up with 0.56.0

Signed-off-by: Christian Nunciato <chris@nunciato.org>

reset committed:
Switch to `prost` from `protobuf` crate

With this commit we say goodbye to the `protobuf` crate that we've been
using for years. The ergonomics of the crate was a bit cumbersome and
there was a severe lack of documentation, but the real reason to switch
here is follows:

* The code generator for the `protobuf` crate had no way to inject our
  own logic into the generation process. This meant that each type
  generated could not be decorated but instead had to be wrapped using
  the new-type pattern. This added a lot of overhead for boiler plate
  protocol level code and was the cause of many inconsistencies across
  our code base. Some crates always wrap, some never do, and some only
  do when it's strictly necessary.
* Generated code would create accessors for getting and setting fields.
  all fields of the generated structs were marked private. This would
  mean that implementing things like Serialization and Deserialization
  would need to be done by hand both because you are unable to decorate
  the generated structs with your own `derive`s but also because the
  fields were not all used to store basic data.
* The generated accessors would provide a default value for all optional
  fields. This meant that you may be overlooking a protocol
  inconsistency which was meant to be handled by the application and
  instead would return a completely false default value that the
  developer never intended.
* Generated fields would include non-std Rust types and the gymnastics
  around getting a simple vector into a field was not only a pain but
  also required application code to know about the protobuf library.
  Ideally the protocol crate and client code are the only crates which
  have a dependency on `protobuf`/`prost`.

There are a lot of improvements here by switching to `prost` but the
developer still needs to take care to make sure that the Supervisor is
robust enough to handle messages from older clients and clients need to
degrade if they receive a message from a newer Supervisor. The
prost-generated code should make this more clear to the developer.

Signed-off-by: Jamie Winsor <jamie@vialstudios.com>

krnowak committed:
pkg-export-k8s: Make some help strings consistent

Should start with uppercase, just like others.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

christophermaier committed:
Minor wording tweaks

Signed-off-by: Christopher Maier <cmaier@chef.io>

christophermaier committed:
0.56.0 Release Blog Post

Signed-off-by: Christopher Maier <cmaier@chef.io>

reset committed:
Bump to Rust 1.26.1

Signed-off-by: Jamie Winsor <jamie@vialstudios.com>

christophermaier committed:
Use octal for permissions of CTL_SECRET!

Signed-off-by: Christopher Maier <cmaier@chef.io>

reset committed:
Merge pull request #5129 from habitat-sh/cm/fix-ctl-secret-permissions

Use octal for permissions of CTL_SECRET!

reset committed:
Merge pull request #5126 from habitat-sh/prost-sup

Switch to `prost` from `protobuf` crate

christophermaier committed:
Revert "Merge pull request #5126 from habitat-sh/prost-sup"

This reverts commit 2c0d9f0bc570a7ec497593e6ed3ef282b7833ae0, reversing
changes made to 251fe99ff3a23b8f4c51220ffb681f2cd99360ce.

Signed-off-by: Christopher Maier <cmaier@chef.io>

reset committed:
Merge pull request #5136 from habitat-sh/cm/revert-prost

Revert "Merge pull request #5126 from habitat-sh/prost-sup"

reset committed:
Revert "Revert "Merge pull request #5126 from habitat-sh/prost-sup""

This reverts commit 93b2d8e6b389de14a2f703a24832337f54345dd8.

christophermaier committed:
Remove unused import

Signed-off-by: Christopher Maier <cmaier@chef.io>

raskchanky committed:
Merge pull request #5137 from habitat-sh/cm/remove-warnings

Remove unused import

reset committed:
Run sup-protocol and sup-client test suites on build server

Signed-off-by: Jamie Winsor <jamie@vialstudios.com>

reset committed:
Write a proper default value for group on spec creation

The default value of a group will now be used when writing a spec file
instead of a blank string

Signed-off-by: Jamie Winsor <jamie@vialstudios.com>

christophermaier committed:
Update documentation for CLI and template data

Signed-off-by: Christopher Maier <cmaier@chef.io>

christophermaier committed:
Merge pull request #5114 from habitat-sh/cdn/sup-tuts

Update Supervisor Documentation and Tutorials for 0.56.0

christophermaier committed:
Bump to 0.57.0-dev

Signed-off-by: Christopher Maier <cmaier@chef.io>

christophermaier committed:
Merge pull request #5139 from habitat-sh/cm/bump

Bump to 0.57.0-dev

christophermaier committed:
Merge pull request #4803 from habitat-sh/expeditor-for-changelogs

Initial Expeditor configuration for updating CHANGELOG.md

christophermaier committed:
Merge pull request #5130 from habitat-sh/rust-1-26-1

Bump to Rust 1.26.1

christophermaier committed:
Remove trailing colon!

Oopsie!

Signed-off-by: Christopher Maier <cmaier@chef.io>

christophermaier committed:
Merge pull request #5141 from habitat-sh/cm/tweak-expeditor

Remove trailing colon!

mwrock committed:
add instructions for rerunning choco tests after publishing release

Signed-off-by: mwrock <matt@mattwrock.com>

christophermaier committed:
Merge pull request #5138 from habitat-sh/prost-sup

Switch from the `protobuf` crate to the `prost` crate

reset committed:
Merge pull request #5142 from habitat-sh/choco_release

add instructions for rerunning choco tests after publishing release

Commit:
Update CHANGELOG.md with details from pull request #5142

krnowak committed:
pkg-export-*: Run rustfmt

The exporters code was not reformatted when the rust version was
bumped to 1.26.1, which also updated rustfmt.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
pkg-export-*: Use failure-related crates from crates registry

In the beginning I specified the git field to point to the original
author and the version constraint to "~1" in Cargo.toml, but then I
found out that some prost package uses failure from crates registry,
so I opted for downgrading to just use a single version of the crates.

On top of that we were pulling failure crates from the git repo of the
original author. But now failure is a part of rust lang nursery. I
expect that the code from nursery gets packaged and pushed into crates
registry. Admittedly for some reason, the newest version available
from git wasn't yet pushed to the registry.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>
